### PR TITLE
Use `int64_t` for nll_loss with cuda inputs

### PR DESCRIPTION
**Pull request by crcrpar**
Related #85005

### CODE REVIEW

File: aten/src/ATen/native/cuda/Loss.cu [hunk 1]
```diff
@@ -445,17 +443,18 @@ __global__ void nll_loss_backward_reduce_cuda_kernel_2d(
...
-    int n_classes,
+    int64_t n_classes,
     int64_t ignore_index) {
   const auto grad = -(size_average ? *grad_output / *total_weight
                                    : *grad_output);
 
   for (int i = threadIdx.x; i < nframe; i += NLL_LOSS_THREADS) {
-    int t = target[i];
-    if (t != static_cast<int>(ignore_index)) {
+    const int64_t t = static_cast<int64_t>(target[i]);
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Loss.cu:452]: There's no real need to make the cast explicit (same on the other functions), but whatever works.

File: aten/src/ATen/native/cuda/Loss.cu [hunk 2]
```diff
@@ -445,17 +443,18 @@ __global__ void nll_loss_backward_reduce_cuda_kernel_2d(
...
+    if (t != ignore_index) {
       CUDA_KERNEL_ASSERT(t >= 0 && t < n_classes);
-      grad_input[i * ndim + t] = weights != nullptr ? weights[t] * grad
-                                                    : grad;
+      const uint64_t index = static_cast<uint64_t>(i) * ndim + t;
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Loss.cu:455]: Let's use `int64_t`  or `uint64_t` to index `input` and `grad_input`, but let's be consistent.

File: aten/src/ATen/native/cuda/Loss.cu [hunk 3]
```diff
@@ -422,16 +422,14 @@ __global__ void nll_loss_backward_reduce_cuda_kernel_1d(
...
   scalar_t *total_weight,
   bool size_average,
-  int n_classes,
+  int64_t n_classes,
   int64_t ignore_index
 ) {
-  int t = static_cast<int>(*target);
-  if (t != static_cast<int>(ignore_index)) {
+  int64_t t = static_cast<int64_t>(*target);
+  if (t != static_cast<int64_t>(ignore_index)) {
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Loss.cu:429]: `ignore_index` is already of type `index64_t`.

File: aten/src/ATen/native/cuda/Loss.cu (hunk 2 above)
Comment (lezcano) [reply aten/src/ATen/native/cuda/Loss.cu:455]: Also, in all the other kernels we have not cast `i`, so I don't think it's necessary to do so here either. You can simply do
```c++
const auto index = i * ndim + t;
```

**Review by lezcano — COMMENTED** [review]
The implementation of this monster of a function is also located at `NLLLoss2d.cu`. Could you apply similar changes to that branch, and add the relevant tests to `test_nn.py`?

**Comment by crcrpar** [comment]
> The implementation of this monster of a function is also located at `NLLLoss2d.cu`. Could you apply similar changes to that branch, and add the relevant tests to `test_nn.py`?

I still don't fully understand how nll_loss function is composed. Why does this function have many variations in multiple files?

**Comment by lezcano** [comment]
> I still don't fully understand how nll_loss function is composed. Why does this function have many variations in multiple files?

Yes it does. ITs implementation is an absolute mess spanning over 2.5k LOC.

File: aten/src/ATen/native/cuda/Loss.cu [hunk 4]
```diff
@@ -187,11 +187,11 @@ __global__ void nll_loss_forward_reduce_cuda_kernel_1d(
...
     scalar_t* weights,
     bool size_average,
-    int n_classes,
+    int64_t n_classes,
     int64_t ignore_index) {
   CUDA_KERNEL_ASSERT(threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0);
 
-  int t = static_cast<int>(*target);
+  int64_t t = static_cast<int64_t>(*target);
   if (t != static_cast<int>(ignore_index)) {
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Loss.cu:195]: I would grep for `int[^36]` in these files and audit all the uses of `int`s in these files to make sure that's consistent.

**Review by lezcano — COMMENTED** [review]
I think the best approach would be to figure out what variables should be of what types, and check every kernel for consistency. At the moment it's not clear whether `i` * ndim should be of type `int64_t` or `int`. Same for the other indexing / values.

Also, even better it may be to simply use `index_t` everywhere, as it should be the right size I think.

**Comment by crcrpar** [comment]
> Also, even better it may be to simply use `index_t` everywhere, as it should be the right size I think.

I don't think it'll work unless we start refactoring of nll loss implementation because `index_t` is either Long or Bool:
https://github.com/pytorch/pytorch/blob/db40fbdee03920944219588464d38774ca0b3d05/aten/src/ATen/native/cuda/Loss.cu#L155-L158

**Comment by lezcano** [comment]
oof, it even has its own index dispatcher. What an absolute mess of a function.

**Comment by crcrpar** [comment]
> The implementation of this monster of a function is also located at `NLLLoss2d.cu`. Could you apply similar changes to that branch, and add the relevant tests to `test_nn.py`?

From what I understand, what `NLLLoss2d.cu` implements are consumed by ATen/native/LossNLL2d.cpp to define `nll_loss_2d`. Then ATen/native/LossNLL.cpp defines `nll_loss_nd` which calls nll_loss_2d inside of it. I didn't get how to correctly add a test case for NLLLoss2d.cu implementations.
Do you have any idea?

**Comment by lezcano** [comment]
Can't you test ti through `nll_loss_nd`, as it calls into it?

**Comment by crcrpar** [comment]
then I think it's already covered as `nll_loss_nd` is what `torch.nn.functional.nll_loss` calls: https://github.com/pytorch/pytorch/blob/291b080e8c54cdb19dd823728cf30fffece10f4d/torch/nn/functional.py#L2697

**Comment by lezcano** [comment]
So, at the moment, the test calls `nll_loss` once, so just one of the implementations must be covered by the test, right? We would need a second call in the test to cover both the implementations.

**Comment by t-vi** [comment]
> oof, it even has its own index dispatcher. What an absolute mess of a function.

Well, so it has the index dispatcher for different dtypes of the class input, which does not seem unreasonable (the alternative being to always cast the index tensor before running the kernel).

But even with this, it seems people want to have tensor indices in 64 bit.
I'm just a bit curious about the perf impact along these lines:

- generally 64-bit indexing comes with quite a penalty,
- on the other hand adding 32 and 64 bit variants is not that attractive, either given that we already have 3 * number of floating types * number of allowed class index types variants.
- The perf impact might be not that grave for nll loss, as it is not a very dense operation.

**Review by t-vi — APPROVED** [review]
To my mind, the patch is good, with maybe checking that a more common case (say 1024 classes, batch size 1024) does not get overly slow (either because the relative impact is small or the absolute time is not large compared to what you do to compute the loss inputs).

Thank you @crcrpar for working on this.

File: aten/src/ATen/native/cuda/Loss.cu (hunk 2 above)
Comment (t-vi) [reply aten/src/ATen/native/cuda/Loss.cu:456]: I think we typically use int64_t for indexing and name the integer types rather than auto in PyTorch code, but that is just my observation.

**Comment by lezcano** [comment]
> Well, so it has the index dispatcher for different dtypes of the class input, which does not seem unreasonable (the alternative being to always cast the index tensor before running the kernel).

My point is, why not using `AT_DISPATCH_INDEX_TYPES`? If we used that, we wouldn't need to bake in `int64_t` or `int`, we could simply use `index_t` where appropriate and `auto` to deduce the correct type. This way we would keep the speed of `int32_t` whenever `target` is of dtype `int32`.

**Comment by t-vi** [comment]
> My point is, why not using AT_DISPATCH_INDEX_TYPES? If we used that, we wouldn't need to bake in int64_t or int, we could simply use index_t where appropriate and auto to deduce the correct type. This way we would keep the speed of int32_t whenever target is of dtype int32.

Right, if we renamed the current `index_t` in that file to `class_index_t` or somesuch, one could use `index_t`. 
I am not sure that `AT_DISPATCH_INDEX_TYPES` is the right thing here. I could be wrong, but at first glance it is about dtypes of inputs, which is not the thing we need here. For the type of change we have here, I am always looking at `cuda::detail::canUse32BitIndexMath` to switching between 32 and 64 bit indexing.

That said, between 1d/2d/nd variants for the kernel and the current dispatching (for floating dtype of the logprob input and class index input dtype), we would multiply the number of kernels by another factor of two. We used to not want to have too many kernels, I'm not sure if I am just being outdated here or whether this still is a consideration.

**Comment by lezcano** [comment]
You are right. There are two variables here`C` and the indexing into `input` (assuming `input` of size `[N, C]`). 

`C` controls the values that may appear within `target`. The precondition here would be "target must be of type `int64` if `C` is larger than 2^31-1. Same applies for `n_classes` and `ignore_index`. Note that now we just allow `target` to have dtype `int64`. This means that those two variables should *always* be of type `int64_t`.

The other one is the one that's causing issues here, and that, as @t-vi suggested, can be controlled via `canUse32BitIndexMath`. Now, we do not have different dispatches for int32 and int64 either, so the way to fix this one would be to audit all the operations and make sure that all the intermediate values that are used to index into `input` are of type `int64_t`.

At the moment, there are still a number of casts to `int` and and a number of these variables that still have type `int`, so those should be changed.

As a side note, I think that `AT_DISPATCH_NLL_LOSS_INDEX_TYPES` is not necessary as it's just used on part of the implementation. In particular, it's not used in `aten/src/ATen/native/cuda/NLLLoss2d.cu`. Also, I bet it's not used in the CPU part of the implementation nor documented. As such, I think it would be safe to simply use `int64_t` for the type of target and remove that weird dispatch,, but this can be done in other PR.

**Review by lezcano — APPROVED** [review]
I'm approving this in the interest of everyone's sanity, as this function and the way it's currently implemented is absolute madness.

Leaving ther ethe point that, for the record, there are still a number of incorrect implicit castings from int64_t to in, like in line 173, `int cur_target = target[index];`. There are also a number of intermediate values that are incorrectly computed in `int`, and values like `n_classes`  should also be `int64_t`s and they are currently being passed around as ints..
